build/enso-formatter/src/main.rs

```diff
@@ -0,0 +1,600 @@
+    token:     HeaderToken,
+    reg_match: String,
+}
+
+impl Debug for HeaderElement {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}({:?})", self.token, self.reg_match.as_str())
+    }
+}
+
+impl HeaderElement {
+    /// Constructor.
+    pub fn new(token: HeaderToken, reg_match: String) -> Self {
+        let attrs = Default::default();
+        Self { attrs, token, reg_match }
+    }
+
+    /// Check whether the element is empty.
+    pub fn is_empty(&self) -> bool {
+        self.reg_match.is_empty()
+    }
+
+    /// Length of the splice. Includes the length of the matched string and all attached attributes.
+    pub fn len(&self) -> usize {
+        let args_len: usize = self.attrs.iter().map(|t| t.len()).sum();
+        self.reg_match.len() + args_len
+    }
+
+    /// Convert the element to a string representation.
+    #[allow(clippy::inherent_to_string)]
+    pub fn to_string(&self) -> String {
+        format!("{}{}", self.attrs.join(""), self.reg_match)
+    }
+}
+
+/// Wrappers for [`Regex::find`] which returns [`Result::Err`] on element found. It allows combining
+/// multiple calls to this function with the Rust `?` syntax.
+fn find_with<T>(input: &str, regex: &Regex, f: impl FnOnce(String) -> T) -> Result<(), T> {
+    match regex.find(input) {
+        Some(t) => Err(f(t.as_str().into())),
+        None => Ok(()),
+    }
+}
+
+/// Regex constructor that starts on the beginning of a line, can be surrounded by whitespaces and
+/// ends with a line break.
+fn re(input: &str) -> Regex {
+    let str = format!(r"^ *{} *(; *)?((\r\n?)|\n)", input);
+    Regex::new(&str).unwrap()
+}
+
+macro_rules! define_rules {
+    ($($name:ident = $re:tt;)*) => {
+        #[allow(non_upper_case_globals)]
+        mod static_re {
+            use super::*;
+            lazy_static! {
+                $(
+                    pub static ref $name: Regex = re($re);
+                )*
+            }
+        }
+
+        fn match_header_internal(input: &str) -> Result<(), HeaderElement> {
+            $( find_with(input, &static_re::$name, |t| HeaderElement::new($name, t))?; )*
+            Ok(())
+        }
+    };
+}
+
+define_rules! {
+    EmptyLine            = r"";
+    ModuleDoc            = r"//![^\n\r]*";
+    Comment              = r"//[^\n\r]*";
+    CrateUse             = r"use +crate( *:: *[\w]+)*( +as +[\w]+)?";
+    CrateUseStar         = r"use +crate( *:: *[\w*]+)*";
+    CratePubUse          = r"pub +use +crate( *:: *[\w]+)*( +as +[\w]+)?";
+    CratePubUseStar      = r"pub +use +crate( *:: *[\w*]+)*";
+    Use                  = r"use +[\w]+( *:: *[\w]+)*( +as +[\w]+)?";
+    UseStar              = r"use +[\w]+( *:: *[\w*]+)*";
+    PubUse               = r"pub +use +[\w]+( *:: *[\w]+)*( +as +[\w]+)?";
+    PubUseStar           = r"pub +use +[\w]+( *:: *[\w*]+)*";
+    ModuleAttribFeature  = r"#!\[feature[^\]]*\]";
+    ModuleAttribFeature2 = r"#!\[allow\(incomplete_features\)\]";
```

Comment:
I think this should be just `ModuleAttribAllow` case.
If for some reason this needs to be separate, it should be named better.